### PR TITLE
Add CRD method to get OutputShape

### PR DESCRIPTION
**Pull request by kumargauravsharma**
Issue #, if available:

Description of changes:

This PR also adds a method `GetOutputShape` to crd which provides output shape

For example, `CreateReplicationGroupOutput`, `ModifyReplicationGroupOutput` wrap `ReplicationGroup` object.
This method is useful in creating reusable code which populates the fields from wrapped object to `ko`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by kumargauravsharma** [comment]
Related PR: aws-controllers-k8s/elasticache-controller#36 where dynamic hook template is used to auto generate reusable setOutput method.

File: pkg/model/crd.go [hunk 1]
```diff
@@ -387,6 +387,34 @@ func (r *CRD) GetOutputWrapperFieldPath(
...
+	// shape and that member shape is a structure, unwrap it.
+	if outputShape.UsedAsOutput && len(outputShape.MemberRefs) == 1 {
+		for _, memberRef := range outputShape.MemberRefs {
+			if memberRef.Shape.Type == "structure" {
+				outputShape = memberRef.Shape
+				break
+			}
+		}
+	}
+	return &outputShape.ShapeName
```
Comment (A-Hilaly) [inline pkg/model/crd.go:415]: what if `memberRef.Shape.Type` is not "structure" ? L415 will panic
Maybe we can return an error when `outputShape` is not found ?

File: pkg/model/crd.go [hunk 2]
```diff
@@ -387,6 +387,34 @@ func (r *CRD) GetOutputWrapperFieldPath(
...
+	}
+
+	// We might be in a "wrapper" shape. Unwrap it to find the real object
+	// representation for the CRD's op. If there is a single member
+	// shape and that member shape is a structure, unwrap it.
+	if outputShape.UsedAsOutput && len(outputShape.MemberRefs) == 1 {
+		for _, memberRef := range outputShape.MemberRefs {
+			if memberRef.Shape.Type == "structure" {
+				outputShape = memberRef.Shape
+				break
```
Comment (A-Hilaly) [inline pkg/model/crd.go:411]: nit: unnecessary break

File: templates/pkg/resource/sdk.go.tpl [hunk 3]
```diff
@@ -335,4 +335,7 @@ func (rm *resourceManager) handleImmutableFieldsChangedCondition(
 
 	return &resource{ko}
 }
-{{- end }}
\ No newline at end of file
+{{- end }}
+{{- if $hookCode := Hook .CRD "sdk_addon" }}
+{{ $hookCode }}
+{{- end }}
```
Comment (A-Hilaly) [inline templates/pkg/resource/sdk.go.tpl:341]: Can you put this in a separate PR or commit ?

File: pkg/model/crd.go [hunk 4]
```diff
@@ -387,6 +387,34 @@ func (r *CRD) GetOutputWrapperFieldPath(
 	return &opConfig.OutputWrapperFieldPath
 }
 
+// OutputShapeName returns name of the Output shape for given operation.
+func (r *CRD) UnwrappedOutputShapeName(
+	// The operation to look for the Output shape
+	op *awssdkmodel.Operation,
+) *string {
```
Comment (A-Hilaly) [inline pkg/model/crd.go:394]: Thinking more about this, i believe we should return the `outputShape` instead of the `outputShape.Name`. This way we can reuse this function other places as well like: https://github.com/aws-controllers-k8s/code-generator/blob/main/pkg/generate/code/set_resource.go#L138-L145

File: pkg/model/crd.go [hunk 5]
```diff
@@ -387,6 +387,34 @@ func (r *CRD) GetOutputWrapperFieldPath(
 	return &opConfig.OutputWrapperFieldPath
 }
 
+// OutputShapeName returns name of the Output shape for given operation.
```
Comment (jaypipes) [inline pkg/model/crd.go:390]: mismatched function name here... plus, I agree with @A-Hilaly about making the function return a Shape instead of a string name. Maybe call this method `GetOutputShape()`?

File: pkg/model/crd.go (hunk 1 above)
Comment (jaypipes) [reply pkg/model/crd.go:415]: > what if `memberRef.Shape.Type` is not "structure" ? L415 will panic
> Maybe we can return an error when `outputShape` is not found ?

Agreed. This should be `return nil` and the above line 410 should be `return memberRef.Shape`

File: templates/pkg/resource/sdk.go.tpl (hunk 3 above)
Comment (jaypipes) [reply templates/pkg/resource/sdk.go.tpl:341]: agree with @A-Hilaly this should go in a separate PR (and I'm not sold on the idea of having a generic "sdk_addon" here. The hooks code was meant for code injection in specific places in the processing pipeline, not really as a way to add custom code outside of the processing pipeline...

**Review by jaypipes — CHANGES_REQUESTED** [review]

File: pkg/model/crd.go (hunk 1 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Thanks for the comments, I have updated the code.
I referred latest set_resource.go logic and moved reusable method into CRD.

File: pkg/model/crd.go (hunk 2 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Updated the code.

File: pkg/model/crd.go (hunk 4 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Changed the return type. However, did not update the set_resource.go as that also sets `sourceVarName` var thus would need to keep its logic.

File: pkg/model/crd.go (hunk 5 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Updated code.

File: templates/pkg/resource/sdk.go.tpl (hunk 3 above)
Comment (kumargauravsharma) [reply templates/pkg/resource/sdk.go.tpl]: Created separate PR for it: https://github.com/aws-controllers-k8s/code-generator/pull/94

File: pkg/model/crd.go [hunk 6]
```diff
@@ -387,6 +387,82 @@ func (r *CRD) GetOutputWrapperFieldPath(
...
+	// representation for the CRD's createOp/DescribeOP.
+
+	// Use the wrapper field path if it's given in the ack-generate config file.
+	wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
+	if wrapperFieldPath != nil {
+		wrapperOutputShape, err := r.GetWrapperOutputShape(outputShape, *wrapperFieldPath)
+		if err != nil {
+			msg := fmt.Sprintf("Unable to unwrap the output shape: %v", err)
+			panic(msg)
+		}
```
Comment (A-Hilaly) [inline pkg/model/crd.go:414]: Can we wrap and return an error instead? All the panic calls are made inside code generators functions - IMO CRD methods should just send simple errors

Comment (kumargauravsharma) [reply pkg/model/crd.go]: Updated the code. Code now returns `nil`.

Comment (A-Hilaly) [reply pkg/model/crd.go]: Why does it return `nil` @kumargauravsharma ? Unless i'm missing something i think it should just return a wrapped error like: `fmt.Errorf("unable to unwrap the output shape: %v", err)`

Comment (A-Hilaly) [reply pkg/model/crd.go]: Ohhh i'm missing the fact that this function doesn't return an error.. I think that it would be nice to return an error anyways for debuging purposes

Comment (kumargauravsharma) [reply pkg/model/crd.go]: Changed method return type to include error as well.

File: pkg/model/crd.go [hunk 7]
```diff
@@ -387,6 +388,81 @@ func (r *CRD) GetOutputWrapperFieldPath(
...
 }
 
+// GetOutputShape returns the Output shape for given operation.
+func (r *CRD) GetOutputShape(
+	// The operation to look for the Output shape
+	op *awssdkmodel.Operation,
+) (*awssdkmodel.Shape, error) {
+	if op == nil {
+		return nil, nil
+	}
```
Comment (A-Hilaly) [inline pkg/model/crd.go:398]: We don't need this block - this means that even if `err` is nil we should perform a "nillity" test on the shape as well. We just have to make sure that either we return and nil `error` or a nil `shape` but never both at the same in the logic bellow.

File: pkg/model/crd.go [hunk 8]
```diff
@@ -387,6 +388,81 @@ func (r *CRD) GetOutputWrapperFieldPath(
...
+
+	// We might be in a "wrapper" shape. Unwrap it to find the real object
+	// representation for the CRD's createOp/DescribeOP.
+
+	// Use the wrapper field path if it's given in the ack-generate config file.
+	wrapperFieldPath := r.GetOutputWrapperFieldPath(op)
+	if wrapperFieldPath != nil {
+		wrapperOutputShape, err := r.GetWrapperOutputShape(outputShape, *wrapperFieldPath)
+		if err != nil {
+			return nil, err
```
Comment (A-Hilaly) [inline pkg/model/crd.go:413]:
```suggestion
			return nil, fmt.Errorf("unable to unwrap the output shape: %v", err)
```

File: pkg/model/crd.go (hunk 7 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Thanks for the comment, I have updated the code.

File: pkg/model/crd.go (hunk 8 above)
Comment (kumargauravsharma) [reply pkg/model/crd.go]: Thanks for the comment, I have updated the code.

**Review by A-Hilaly — APPROVED** [review]
NIce, thanks @kumargauravsharma !:+1:
